Leiningen creates non-backwards compatible data_readers.clj

Steps to reproduce:
```
% docker run -it clojure:openjdk-8-lein bash 
root@aa41a92991df:/tmp# echo '(defproject bug "bug" :dependencies [[org.clojure/clojure "1.8.0"] [org.flatland/ordered "1.5.6"]])' > project.clj
root@aa41a92991df:/tmp# lein uberjar
Created /tmp/target/bug-bug.jar
Created /tmp/target/bug-bug-standalone.jar
root@aa41a92991df:/tmp# java -cp /tmp/target/bug-bug-standalone.jar clojure.main
clojure.lang.LispReader$ReaderException: java.lang.RuntimeException: Reader tag must be a symbol
	at clojure.lang.LispReader.read(LispReader.java:292)
	at clojure.lang.LispReader.read(LispReader.java:196)
	at clojure.lang.LispReader.read(LispReader.java:185)
	at clojure.core$read.invokeStatic(core.clj:3666)
	at clojure.core$load_data_reader_file.invokeStatic(core.clj:7485)
	at clojure.core$load_data_reader_file.invoke(core.clj:7480)
	at clojure.core.protocols$fn__6755.invokeStatic(protocols.clj:167)
	at clojure.core.protocols$fn__6755.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__6710$G__6705__6719.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__6738.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__6738.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__6684$G__6679__6697.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6545)
	at clojure.core$load_data_readers$fn__7291.invoke(core.clj:7511)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.Var.alterRoot(Var.java:303)
	at clojure.core$alter_var_root.invokeStatic(core.clj:5299)
	at clojure.core$alter_var_root.doInvoke(core.clj:5294)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at clojure.core$load_data_readers.invokeStatic(core.clj:7510)
	at clojure.core$fn__7294.invokeStatic(core.clj:7515)
	at clojure.core$fn__7294.invoke(core.clj:7515)
	at clojure.core__init.load(Unknown Source)
	at clojure.core__init.<clinit>(Unknown Source)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at clojure.lang.RT.classForName(RT.java:2168)
	at clojure.lang.RT.classForName(RT.java:2177)
	at clojure.lang.RT.loadClassForName(RT.java:2196)
	at clojure.lang.RT.load(RT.java:443)
	at clojure.lang.RT.load(RT.java:419)
	at clojure.lang.RT.doInit(RT.java:461)
	at clojure.lang.RT.<clinit>(RT.java:331)
	at clojure.main.<clinit>(main.java:20)
Caused by: java.lang.RuntimeException: Reader tag must be a symbol
	at clojure.lang.LispReader$CtorReader.invoke(LispReader.java:1221)
	at clojure.lang.LispReader$DispatchReader.invoke(LispReader.java:684)
	at clojure.lang.LispReader.read(LispReader.java:263)
	... 35 more
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at clojure.lang.RT.classForName(RT.java:2168)
	at clojure.lang.RT.classForName(RT.java:2177)
	at clojure.lang.RT.loadClassForName(RT.java:2196)
	at clojure.lang.RT.load(RT.java:443)
	at clojure.lang.RT.load(RT.java:419)
	at clojure.lang.RT.doInit(RT.java:461)
	at clojure.lang.RT.<clinit>(RT.java:331)
	at clojure.main.<clinit>(main.java:20)
Caused by: clojure.lang.LispReader$ReaderException: java.lang.RuntimeException: Reader tag must be a symbol
	at clojure.lang.LispReader.read(LispReader.java:292)
	at clojure.lang.LispReader.read(LispReader.java:196)
	at clojure.lang.LispReader.read(LispReader.java:185)
	at clojure.core$read.invokeStatic(core.clj:3666)
	at clojure.core$load_data_reader_file.invokeStatic(core.clj:7485)
	at clojure.core$load_data_reader_file.invoke(core.clj:7480)
	at clojure.core.protocols$fn__6755.invokeStatic(protocols.clj:167)
	at clojure.core.protocols$fn__6755.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__6710$G__6705__6719.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__6738.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__6738.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__6684$G__6679__6697.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6545)
	at clojure.core$load_data_readers$fn__7291.invoke(core.clj:7511)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.Var.alterRoot(Var.java:303)
	at clojure.core$alter_var_root.invokeStatic(core.clj:5299)
	at clojure.core$alter_var_root.doInvoke(core.clj:5294)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at clojure.core$load_data_readers.invokeStatic(core.clj:7510)
	at clojure.core$fn__7294.invokeStatic(core.clj:7515)
	at clojure.core$fn__7294.invoke(core.clj:7515)
	at clojure.core__init.load(Unknown Source)
	at clojure.core__init.<clinit>(Unknown Source)
	... 10 more
Caused by: java.lang.RuntimeException: Reader tag must be a symbol
	at clojure.lang.LispReader$CtorReader.invoke(LispReader.java:1221)
	at clojure.lang.LispReader$DispatchReader.invoke(LispReader.java:684)
	at clojure.lang.LispReader.read(LispReader.java:263)
	... 35 more
```

This is because the generated `data_readers.clj` uses namespaced map syntax, which clojure 1.8 doesn't support:

```
root@aa41a92991df:/tmp# unzip -c /tmp/target/bug-bug-standalone.jar data_readers.clj
Archive:  /tmp/target/bug-bug-standalone.jar
  inflating: data_readers.clj        
#:ordered{set flatland.ordered.set/into-ordered-set, map flatland.ordered.map/ordered-map}

--- a/src/leiningen/uberjar.clj
+++ b/src/leiningen/uberjar.clj
@@ -53,10 +53,16 @@
   "Project `:uberjar-merge-with` merger for components.xml files."
   [components-read into components-write])
 
+(defn- pr-str-backwards-compatibly
+  "Produce EDN that should work with older clojure versions."
+  [val]
+  (binding [*print-namespace-maps* false]
+    (pr-str val)))
+
 (def clj-map-merger
   "Project `:uberjar-merge-with` for files containing a single map
   read with `clojure.core/read`, such as data_readers.clj."
-  [(comp read-string slurp) merge #(spit %1 (pr-str %2))])
+  [(comp read-string slurp) merge #(spit %1 (pr-str-backwards-compatibly %2))])
 
 (defn- merger-match? [[pattern] filename]
   (boolean

--- a/test/leiningen/test/helper.clj
+++ b/test/leiningen/test/helper.clj
@@ -62,6 +62,8 @@
 
 (def uberjar-merging-project (read-test-project "uberjar-merging"))
 
+(def data-readers-backwards-compatibility-project (read-test-project "data-readers-backwards-compatibility"))
+
 (def overlapped-sourcepaths-project (read-test-project "overlapped-sourcepaths"))
 
 (def more-gen-classes-project (read-test-project "more-gen-classes"))

--- a/test/leiningen/test/uberjar.clj
+++ b/test/leiningen/test/uberjar.clj
@@ -7,6 +7,7 @@
             [leiningen.test.helper :refer [unmemoize
                                            sample-no-aot-project
                                            uberjar-merging-project
+                                           data-readers-backwards-compatibility-project
                                            provided-project
                                            managed-deps-project
                                            managed-deps-snapshot-project]])
@@ -47,6 +48,23 @@
                (->> (.getEntry zf "data_readers.clj")
                     (.getInputStream zf)
                     slurp read-string)))))))
+
+(deftest test-uberjar-data-readers-backwards-compatibility
+  (uberjar data-readers-backwards-compatibility-project)
+  (let [filename (str "test_projects/data-readers-backwards-compatibility/"
+                      "target/bug-bug-standalone.jar")
+        uberjar-file (File. filename)]
+    (is (= true (.exists uberjar-file)))
+    (when (.exists uberjar-file)
+      (.deleteOnExit uberjar-file)
+      (with-open [zf (ZipFile. uberjar-file)]
+        (let [contents (->> (.getEntry zf "data_readers.clj")
+                            (.getInputStream zf)
+                            slurp)]
+          (is (.startsWith contents "{")) ;; not a namespaced map
+          (is (= '{ordered/set flatland.ordered.set/into-ordered-set
+                   ordered/map flatland.ordered.map/ordered-map}
+                 (read-string contents))))))))
 
 (deftest test-components-merger
   (let [file1 (str "test_projects/uberjar-components-merging/components1.xml")

--- a/test_projects/data-readers-backwards-compatibility/project.clj
+++ b/test_projects/data-readers-backwards-compatibility/project.clj
@@ -1,0 +1,8 @@
+;; This project is used for leiningen's test suite, so don't change
+;; any of these values without updating the relevant tests. If you
+;; just want a basic project to work from, generate a new one with
+;; "lein new".
+
+(defproject bug "bug"
+  :dependencies [[org.clojure/clojure "1.8.0"]
+                 [org.flatland/ordered "1.5.6"]])

